Modules/_copy.c

```diff
@@ -0,0 +1,437 @@
+{
+    unsigned i;
+    PyObject* y, * id_x;
+    const struct deepcopy_dispatcher* dd;
+    Py_hash_t hash_id_x;
+
+    assert(PyDict_CheckExact(memo));
+
+    PyTypeObject *type_x = Py_TYPE(x);
+
+    /*
+     * No need to have a separate dispatch function for this. Also, the
+     * array would have to be quite a lot larger before a smarter data
+     * structure is worthwhile.
+     */
+    for (i = 0; i < N_ATOMIC_TYPES; ++i) {
+        if (type_x == atomic_type[i]) {
+            return Py_NewRef(x);
+        }
+    }
+
+    /* Have we already done a deepcopy of x? */
+    id_x = object_id(x);
+    if (id_x == NULL)
+        return NULL;
+    hash_id_x = PyObject_Hash(id_x);
+
+    y = _PyDict_GetItem_KnownHash(memo, id_x, hash_id_x);
+    if (y != NULL) {
+        Py_DECREF(id_x);
+        Py_INCREF(y);
+        return y;
+    }
+    /*
+     * Hold on to id_x and its hash a little longer - the dispatch handlers
+     * will all need it.
+     */
+    for (i = 0; i < N_DISPATCHERS; ++i) {
+        dd = &deepcopy_dispatch[i];
+        if (type_x != dd->type)
+            continue;
+
+        y = dd->handler(module, x, memo, id_x, hash_id_x);
+        Py_DECREF(id_x);
+        if (y == NULL)
+            return NULL;
+        if (x != y && memo_keepalive(x, memo) < 0) {
+            Py_DECREF(y);
+            return NULL;
+        }
+        return y;
+    }
+
+    Py_DECREF(id_x);
+
+    return do_deepcopy_fallback(module, x, memo);
+}
+
+/*
+ * This is the Python entry point. Hopefully we can stay in the C code
+ * most of the time, but we will occasionally call the Python code to
+ * handle the stuff that's very inconvenient to write in C, and that
+ * will then call back to us.
+ */
+
+/*[clinic input]
+ _copy.deepcopy
+
+   x: object
+   memo: object = None
+   /
+
+ Create a deep copy of x
+
+ See the documentation for the copy module for details.
+
+[clinic start generated code]*/
+
+static PyObject *
+_copy_deepcopy_impl(PyObject *module, PyObject *x, PyObject *memo)
+/*[clinic end generated code: output=825a9c8dd4bfc002 input=519bbb0201ae2a5c]*/
+{
+    PyObject* result;
+
+    if (memo == Py_None) {
+        memo = PyDict_New();
+        if (memo == NULL)
+            return NULL;
+    }
+    else {
+        if (!PyDict_CheckExact(memo)) {
+            PyErr_SetString(PyExc_TypeError, "memo must be a dict");
+            return NULL;
+        }
+        Py_INCREF(memo);
+    }
+
+    result = do_deepcopy(module, x, memo);
+
+    Py_DECREF(memo);
+    return result;
+}
+
+static PyMethodDef copy_functions[] = {
+    _COPY_DEEPCOPY_METHODDEF
+    {NULL, NULL}
+};
+
+static int
+copy_clear(PyObject *module)
+{
+    copy_module_state *state = get_copy_module_state(module);
+    Py_CLEAR(state->python_copy_module);
+    return 0;
+}
+
+static void
+copy_free(void *module)
+{
+    copy_clear((PyObject *)module);
+}
+
+static struct PyModuleDef copy_moduledef = {
```

Comment:
You should use module slots to establish module state and import modules `PyImport_ImportModule`.

See `_bisectmodule.c` for an example.